File: app/models/banner.rb
# typed: strict
# frozen_string_literal: true

# == Schema Information
#
# Table name: banners
#
#  id          :bigint           not null, primary key
#  background  :string           default("cta_gradient"), not null
#  body        :text             not null
#  finishes_at :datetime
#  locale      :string           not null
#  starts_at   :datetime
#  state       :string           default("draft"), not null
#  url         :string
#  created_at  :datetime         not null
#  updated_at  :datetime         not null
#
# Indexes
#
#  index_banners_on_locale_and_state  (locale,state)
#
class Banner < ApplicationRecord
  class State < T::Enum
    enums do
      Draft = new("draft")
      Published = new("published")
      Archived = new("archived")
    end
  end

  class Locale < T::Enum
    enums do
      Ru = new("ru")
      En = new("en")
    end
  end

  class Background < T::Enum
    enums do
      CtaGradient = new("cta_gradient")
      Dark = new("dark")
      Blue = new("blue")
    end
  end

  include BannerRepository

  typed_enum :state, State, suffix: true, validate: true, default: State::Draft
  typed_enum :locale, Locale, suffix: true, validate: true
  typed_enum :background, Background, suffix: true, validate: true, default: Background::CtaGradient

  validates :body, presence: true
  validates :locale, presence: true

  sig { params(_auth_object: T.untyped).returns(T.untyped) }
  def self.ransackable_attributes(_auth_object = nil)
    [ "id", "created_at", "state", "locale", "starts_at", "finishes_at" ]
  end

  sig { params(_auth_object: T.untyped).returns(T.untyped) }
  def self.ransackable_associations(_auth_object = nil)
    []
  end

  sig { returns(String) }
  def to_s
    body.to_s.truncate(50)
  end
end

File: src/components/admin/resources/banner.ts
import { useTranslation } from "react-i18next";
import type { Banner, BannerInput } from "@/client/types.gen";
import type { CrudFieldSpec } from "@/components/admin/CrudForm";

// Banner form values ARE the generated request body (BannerInput): the datetime
// fields carry RFC3339 strings (the engine's DateTimeField hides the Mantine
// picker format), so the form validates against the generated `zBannerInput`
// directly — no hand-written schema, no per-resource date conversion.

export const emptyBanner: BannerInput = {
  state: "draft",
  background: "cta_gradient",
  locale: "en",
  body: "",
  url: "",
  startsAt: null,
  finishesAt: null,
};

// Seed the form from an existing row: pick only the writable fields (drops
// id/createdAt so they never leak into the request body) and keep `url` a string
// so the text input stays controlled.
export function bannerToForm(banner: Banner): BannerInput {
  return {
    state: banner.state,
    background: banner.background,
    locale: banner.locale,
    body: banner.body,
    url: banner.url ?? "",
    startsAt: banner.startsAt,
    finishesAt: banner.finishesAt,
  };
}

// Field descriptors driving CrudForm. Enum option labels reuse the legacy
// `*/values` i18n maps, so admin copy matches the Rails back office.
export function useBannerFields(): CrudFieldSpec<BannerInput>[] {
  const { t } = useTranslation();
  return [
    {
      name: "state",
      label: t(($) => $.models.attributes.banner.state),
      type: "select",
      required: true,
      options: [
        {
          value: "draft",
          label: t(($) => $.models.attributes.banner["state/values"].draft),
        },
        {
          value: "published",
          label: t(($) => $.models.attributes.banner["state/values"].published),
        },
        {
          value: "archived",
          label: t(($) => $.models.attributes.banner["state/values"].archived),
        },
      ],
    },
    {
      name: "background",
      label: t(($) => $.models.attributes.banner.background),
      type: "select",
      required: true,
      options: [
        {
          value: "cta_gradient",
          label: t(($) => $.models.attributes.banner["background/values"].cta_gradient),
        },
        {
          value: "dark",
          label: t(($) => $.models.attributes.banner["background/values"].dark),
        },
        {
          value: "blue",
          label: t(($) => $.models.attributes.banner["background/values"].blue),
        },
      ],
    },
    {
      name: "locale",
      label: t(($) => $.models.attributes.banner.locale),
      type: "select",
      required: true,
      options: [
        {
          value: "en",
          label: t(($) => $.models.attributes.banner["locale/values"].en),
        },
        {
          value: "ru",
          label: t(($) => $.models.attributes.banner["locale/values"].ru),
        },
      ],
    },
    {
      name: "body",
      label: t(($) => $.models.attributes.banner.body),
      type: "textarea",
      required: true,
    },
    {
      name: "url",
      label: t(($) => $.models.attributes.banner.url),
    },
    {
      name: "startsAt",
      label: t(($) => $.models.attributes.banner.starts_at),
      type: "datetime",
    },
    {
      name: "finishesAt",
      label: t(($) => $.models.attributes.banner.finishes_at),
      type: "datetime",
    },
  ];
}
